### ISSUE TITLE
Question concerning 'make'

### ISSUE THREAD

**Issue by benediktahrens** [UniMath/UniMath#341]
When calling
$ make -j2 install world
I get the behaviour shown below, which does not arise when just calling
$ make -j2 install
Can you explain what is going on?

"sub/coq/bin/coqdep" -c -Q "UniMath" UniMath "UniMath/Foundations/Basics/Preamble.v" > "UniMath/Foundations/Basics/Preamble.v.d" || ( RV=$?; rm -f "UniMath/Foundations/Basics/Preamble.v.d"; exit ${RV} )
"sub/coq/bin/coqc"  -q  -Q "UniMath" UniMath  -indices-matter -type-in-type  UniMath/Foundations/Basics/Preamble
"sub/coq/bin/coqc"  -q  -Q "UniMath" UniMath  -indices-matter -type-in-type  UniMath/Foundations/Basics/Preamble
0 = 0
     : UUU
0 = 0
     : UUU
"sub/coq/bin/coqc"  -q  -Q "UniMath" UniMath  -indices-matter -type-in-type  UniMath/Foundations/Basics/PartA
"sub/coq/bin/coqc"  -q  -Q "UniMath" UniMath  -indices-matter -type-in-type  UniMath/Foundations/Basics/PartA

**Comment by DanGrayson** [comment]
Because of the lines
```
$(VOFILES): %.vo: %.v
	$(COQC) $(COQDEBUG) $(COQFLAGS) $*

$(GLOBFILES): %.glob: %.v
	$(COQC) $(COQDEBUG) $(COQFLAGS) $*
```
in the file `build/CoqMakefile.make`.  The `*.glob` files are needed for building the documentation,
and the `*.vo` files are what we make when simply compiling our proofs.

I thought I might be able to work around it by requesting just the `*.glob` files to be built, but `Preamble.vo` is needed to make `PartA.glob`.

I recommend using two make commands: `make ; make world install`.  The first one builds all the *.vo files on purpose and all the *.glob files as a side effect.  The second one finds that the *.glob files are already there.

**Comment by benediktahrens** [comment]
On 04/03/2016 02:25 PM, Daniel R. Grayson wrote:
> Because of the lines
> ```
> $(VOFILES): %.vo: %.v
> 	$(COQC) $(COQDEBUG) $(COQFLAGS) $*
> 
> $(GLOBFILES): %.glob: %.v
> 	$(COQC) $(COQDEBUG) $(COQFLAGS) $*
> ```
> in the file `build/CoqMakefile.make`. The `*.glob` files are needed
for building the documentation,
> and the `*.vo` files are what we make when simply compiling our 
> proofs.
> 
> I thought I might be able to work around it by requesting just the
`*.glob` files to be built, but `Preamble.vo` is needed to make
`PartA.glob`.
> 
> I recommend using two make commands: `make ; make world install`. 
> The
first one builds all the *.vo files on purpose and all the *.glob files
as a side effect. The second one finds that the *.glob files are already
there.
> 

Thanks for the information! I will do as you recommend, i.e., building
the documentation separately.

**Comment by DanGrayson** [comment]
The info file for GNU make says this: "Pattern rules may have more than one target.  Unlike normal rules, this does not act as many different rules with the same prerequisites and recipe.  If a pattern rule has multiple targets, 'make' knows that the rule's recipe is responsible for making all of the targets.  The recipe is executed only once to make all the targets.  When searching for a pattern rule to match a target, the target patterns of a rule other than the one that matches the target in need of a rule are incidental: 'make' worries only about giving a recipe and prerequisites to the file presently in question.  However, when this file's recipe is run, the other targets are marked as having been updated themselves."

Thus the presence of the lines
```
$(VOFILES): %.vo: %.v
    $(COQC) $(COQDEBUG) $(COQFLAGS) $*

$(GLOBFILES): %.glob: %.v
    $(COQC) $(COQDEBUG) $(COQFLAGS) $*
```
in the file made for us by coq_makefile can be regarded as a misfeature of coq_makefile.  It should have instead made this single rule:
```
%.vo %.glob: %.v
    $(COQC) $(COQDEBUG) $(COQFLAGS) $*
```

**Comment by DanGrayson** [comment]
I have submitted this as a bug report for Coq at https://coq.inria.fr/bugs/show_bug.cgi?id=4660

**Comment by JasonGross** [comment]
Actually, the rule used to be
```
%.vo %.glob: %.v
    $(COQC) $(COQDEBUG) $(COQFLAGS) $*
```
This led to much more subtle [bugs to track down](https://coq.inria.fr/bugs/show_bug.cgi?id=3406): if you neglected to list all of necessary `.v` dependencies when running `coq_makefile`, you would sometimes (sporadically) get `Error: Compiled library foo.vo makes inconsistent assumptions over library bar`.

Is there any way of writing the rule to consider `.vo` and `.glob` files at the same time, but still not have any targets for files not in `$(VOFILES)` or `$(GLOBFILES)`?

**Comment by DanGrayson** [comment]
Yes, there is a way:
```
VFILES = aaa.v bbb.v
VOFILES:=$(VFILES:.v=.vo)
GLOBFILES:=$(VFILES:.v=.glob)
VOANDGLOBFILES:=$(VFILES:.v=.vo-and-glob)
all : $(VOFILES) $(GLOBFILES)
$(VOFILES) : %.vo : %.vo-and-glob ; @ true
$(GLOBFILES) : %.glob : %.vo-and-glob ; @ true
$(VOANDGLOBFILES) : %.vo-and-glob : %.v; touch $*.vo $*.glob
.INTERMEDIATE : $(VFILES:.v=.vo-and-glob)
clean :; rm -f aaa.vo aaa.glob
```

**Comment by DanGrayson** [comment]
Reopening, as this is still an issue.

**Comment by DanGrayson** [comment]
My Coq pull request has been open since July: https://github.com/coq/coq/pull/235

**Comment by DanGrayson** [comment]
Hold on, it seems to be fixed in Coq 8.6, as far as I can tell.  (I was running 8.5.)  Closing
